File: Android/Sources/UserControlLibraries/GoogleMapsLib/src/main/java/com/genexus/android/controls/maps/google/GxMapViewCamera.java
package com.genexus.android.controls.maps.google;

import android.graphics.Point;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.genexus.android.core.base.services.Services;
import com.genexus.android.maps.GxMapViewBase;
import com.genexus.android.maps.MapViewCamera;
import com.genexus.android.core.controls.maps.common.IMapLocation;
import com.genexus.android.maps.MapItemViewHelper;
import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;

class GxMapViewCamera extends MapViewCamera<CameraUpdate, MapLocation, MapLocationBounds, GxMapViewItem, GxMapViewData> {
	private final GoogleMap mMap;
	private final MapView mMapView;

	public GxMapViewCamera(GoogleMap map, MapView mapView) {
		mMap = map;
		mMapView = mapView;
	}

	@Override
	public void updateCamera(@NonNull MapLocationBounds bounds, int padding, int duration) {
		CameraUpdate update = CameraUpdateFactory.newLatLngBounds(bounds.getLatLngBounds(), padding);
		updateCamera(update, duration, null);
	}

	@Override
	public void updateCamera(@NonNull IMapLocation location, int zoom, int duration, @Nullable GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		updateCamera(getCameraUpdate(location, zoom), duration, callback);
	}

	@Override
	public void updateCamera(@NonNull IMapLocation location, int zoom) {
		updateCamera(getCameraUpdate(location, zoom), 0, null);
	}

	@Override
	public void updateCamera(CameraUpdate cameraUpdate, int duration, @Nullable GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		if (mMap != null) {
			try {
				if (duration == 0)
					mMap.animateCamera(cameraUpdate, getCallback(callback));
				else
					mMap.animateCamera(cameraUpdate, duration, getCallback(callback));

				return;
			} catch (IllegalStateException e) {
				Services.Log.error(e);
			}
		}

		setPendingUpdate(cameraUpdate);
	}

	@Nullable
	@Override
	public Projection<MapLocation, MapLocationBounds> getProjection(MapLocation initialLocation) {
		if (mMap == null)
			return null;

		com.google.android.gms.maps.Projection projection = mMap.getProjection();
		CameraPosition cameraPosition = mMap.getCameraPosition();

		MapLocation fromScreenLocation = new MapLocation(0D, 0D);
		if (initialLocation != null) {
			Point centerPoint = projection.toScreenLocation(new LatLng(initialLocation.getLatitude(), initialLocation.getLongitude()));
			centerPoint.y -= (int) (mMapView.getHeight() * MapItemViewHelper.MARKER_INFO_WINDOW_OFF_CENTER_FACTOR);
			fromScreenLocation = new MapLocation(projection.fromScreenLocation(centerPoint));
		}

		MapLocation center = new MapLocation(cameraPosition.target.latitude, cameraPosition.target.longitude);
		MapLocationBounds latLngBounds = new MapLocationBounds(projection.getVisibleRegion().latLngBounds);
		int zoom = (int) cameraPosition.zoom;

		return new Projection<>(latLngBounds, center, fromScreenLocation, zoom);
	}

	private CameraUpdate getCameraUpdate(@NonNull IMapLocation location, int zoom) {
		LatLng latLng = new LatLng(location.getLatitude(), location.getLongitude());
		CameraUpdate update;
		if (zoom > 0)
			update = CameraUpdateFactory.newLatLngZoom(latLng, zoom);
		else
			update = CameraUpdateFactory.newLatLng(latLng);

		return update;
	}

	private GoogleMap.CancelableCallback getCallback(GxMapViewBase.OnMarkerCameraUpdateCallback callback) {
		return new GoogleMap.CancelableCallback() {
			@Override
			public void onFinish() {
				if (callback != null)
					callback.onFinish();
			}

			@Override
			public void onCancel() {
				if (callback != null)
					callback.onCancel();
			}
		};
	}
}

File: Android/Sources/UserControlLibraries/MapboxLib/src/main/java/com/genexus/android/controls/maps/mapbox/GxMapViewCamera.kt
package com.genexus.android.controls.maps.mapbox

import com.genexus.android.core.base.services.Services
import com.genexus.android.core.controls.maps.common.IMapLocation
import com.genexus.android.maps.GxMapViewBase
import com.genexus.android.maps.MapItemViewHelper
import com.genexus.android.maps.MapViewCamera
import com.mapbox.mapboxsdk.camera.CameraUpdate
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory
import com.mapbox.mapboxsdk.geometry.LatLng
import com.mapbox.mapboxsdk.maps.MapView
import com.mapbox.mapboxsdk.maps.MapboxMap
import kotlin.math.roundToInt

internal class GxMapViewCamera(private val map: MapboxMap?, private val mapView: MapView) :
    MapViewCamera<CameraUpdate, MapLocation, MapLocationBounds, GxMapViewItem, GxMapViewData>() {

    override fun updateCamera(bounds: MapLocationBounds, padding: Int, duration: Int) {
        val cameraUpdate = CameraUpdateFactory.newLatLngBounds(bounds.latLngBounds, padding)
        updateCamera(cameraUpdate, duration, null)
    }

    override fun updateCamera(location: IMapLocation, zoom: Int) {
        updateCamera(getCameraUpdate(location, zoom), 0, null)
    }

    override fun updateCamera(
        location: IMapLocation,
        zoom: Int,
        duration: Int,
        callback: GxMapViewBase.OnMarkerCameraUpdateCallback?
    ) {
        updateCamera(getCameraUpdate(location, zoom), duration, callback)
    }

    override fun updateCamera(
        cameraUpdate: CameraUpdate,
        duration: Int,
        callback: GxMapViewBase.OnMarkerCameraUpdateCallback?
    ) {
        if (map != null) {
            try {
                if (duration == 0)
                    map.animateCamera(cameraUpdate, getCallback(callback))
                else
                    map.animateCamera(cameraUpdate, duration, getCallback(callback))

                return
            } catch (e: IllegalStateException) {
                Services.Log.error(e)
            }
        }
        pendingUpdate = cameraUpdate
    }

    override fun getProjection(initialLocation: MapLocation?): Projection<MapLocation, MapLocationBounds>? {
        map?.let {
            val projection = it.projection
            val cameraPosition = it.cameraPosition

            var fromScreenLocation = MapLocation(0.toDouble(), 0.toDouble())
            initialLocation?.let {
                val centerPoint = projection.toScreenLocation(LatLng(initialLocation.latitude, initialLocation.longitude))
                centerPoint.y -= (mapView.height * MapItemViewHelper.MARKER_INFO_WINDOW_OFF_CENTER_FACTOR).roundToInt().toFloat()
                fromScreenLocation = MapLocation(projection.fromScreenLocation(centerPoint))
            }

            val center = MapLocation(cameraPosition.target.latitude, cameraPosition.target.longitude)
            val latLngBounds = MapLocationBounds(projection.visibleRegion.latLngBounds)
            val zoom = cameraPosition.zoom.toInt()

            return Projection(latLngBounds, center, fromScreenLocation, zoom)
        }

        return null
    }

    private fun getCameraUpdate(location: IMapLocation, zoom: Int): CameraUpdate {
        val latLng = LatLng(location.latitude, location.longitude)
        return if (zoom > 0)
            CameraUpdateFactory.newLatLngZoom(latLng, zoom.toDouble())
        else
            CameraUpdateFactory.newLatLng(latLng)
    }

    private fun getCallback(callback: GxMapViewBase.OnMarkerCameraUpdateCallback?): MapboxMap.CancelableCallback {
        return object : MapboxMap.CancelableCallback {
            override fun onCancel() {
                callback?.onCancel()
            }

            override fun onFinish() {
                callback?.onFinish()
            }
        }
    }
}
